src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp

```diff
@@ -1087,17 +1116,60 @@ ParallelReadResponse ParallelReplicasReadingCoordinator::handleRequest(ParallelR
 
     ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::ParallelReplicasHandleRequestMicroseconds);
 
-    std::lock_guard lock(mutex);
+    ParallelReadResponse response;
+    response.finish = true;
 
-    if (!pimpl)
-        initialize(request.mode);
+    if (is_reading_completed)
+        return response;
 
-    const auto replica_num = request.replica_num;
-    auto response = pimpl->handleRequest(std::move(request));
-    if (!response.finish)
+    std::set<size_t> replicas_to_exclude;
+    bool reading_assignment_has_been_completed = false;
     {
-        if (replicas_used.insert(replica_num).second)
-            ProfileEvents::increment(ProfileEvents::ParallelReplicasUsedCount);
+        std::lock_guard lock(mutex);
+        if (is_reading_completed)
+            return response;
+
+        if (!pimpl)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Got read request from replica {} without ranges announcement", request.replica_num);
+
+        const auto replica_num = request.replica_num;
+        response = pimpl->handleRequest(std::move(request));
+        if (!response.finish)
+        {
+            chassert(!is_reading_completed);
+
+            if (replicas_used.insert(replica_num).second)
+                ProfileEvents::increment(ProfileEvents::ParallelReplicasUsedCount);
+        }
+        else
+        {
+            if (isReadingCompleted())
+            {
+                reading_assignment_has_been_completed = !is_reading_completed.exchange(true);
+                replicas_to_exclude = replicas_used;
+                // exclude itself from canceling
+                replicas_to_exclude.insert(replica_num);
```

Comment:
i'm not quite getting this logic. could you elaborate it a bit pls? and do we really need to create a local `replicas_to_exclude` that has almost the same content as `replicas_used` per se?

Reply:
This is necessary due to suboptimal pr protocol if replica is taking last reading task.
The current protocol requires additional round-trip to finish reading for a replica, i.e., (1) packet with ranges to read and finish = false and (2) empty packet with finish=true.
So, there is a case when replica N will ask for their first task, but all tasks have already been assigned. It means that replicas N will not be in `replicas_used` set, but it shouldn't be included in replicas for cancelling.